graphql_compiler/compiler/emit_sql.py

```diff
@@ -239,19 +340,152 @@ def start_global_operations(self):
 
     def filter(self, predicate):
         """Execute a Filter Block."""
-        sql_expression = predicate.to_sql(self._aliases, self._current_alias)
+        if self._in_fold or isinstance(predicate.left, FoldedContextField) \
+                or isinstance(predicate.right, FoldedContextField):
+            raise NotImplementedError('Filters inside a fold are not implemented yet.')
+
+        sql_expression = predicate.to_sql(self._aliases,
+                                          self._current_alias,
+                                          self._fold_output_column_aliases)
         if self._is_in_optional_scope():
             sql_expression = sqlalchemy.or_(sql_expression,
                                             self._came_from[self._current_alias].is_(None))
         self._filters.append(sql_expression)
 
+    def fold(self, fold_scope_location):
+        """Begin execution of a Fold Block."""
+        # Subquery will look as follows:
+        #
+        # SELECT
+        #   OuterVertex.uuid <- this value is determined from the edge descriptor
+        #   array_agg(InnerVertex.fold_output_column) AS fold_output
+        # FROM OuterVertex
+        # INNER JOIN ... <- INNER JOINs compiled during Unfold block
+        # ON ...
+        #          ...
+        # INNER JOIN InnerVertex <- INNER JOINs compiled during Unfold block
+        # ON ...
+        # GROUP BY OuterVertex.uuid
+
+        if self._in_fold:
+            raise AssertionError(u'Reached an invalid fold state.')
+
+        # begin the fold
+        self._in_fold = True
+        # outer vertex is whatever vertex occurs immediately before the fold
+        self._fold_outer_vertex_location = self._current_location
+        self._fold_inner_vertex_location = fold_scope_location
+
+        # basic info about the outer vertex and inner vertex
+        inner_vertex_table = self._sql_schema_info.vertex_name_to_table[
+            self._ir.query_metadata_table.get_location_info(fold_scope_location).type.name
+        ].alias()
+        outer_vertex_table = self._current_alias.alias()
+        inner_vertex_type = self._ir.query_metadata_table \
+            .get_location_info(fold_scope_location) \
+            .type \
+            .name
+        outer_vertex_type = self._current_classname
+
+        self._aliases[
+            (self._current_location.query_path, self._fold_inner_vertex_location)
```

Comment:
Not clear to me why we're using the `_current_location.query_path` here. We're assigning the `inner_vertex_table` as an alias for the outer vertex location?

Reply:
I was thinking the keys to `self._aliases` should be `(query_path, fold_path)` (as in the TODO on 186). 

For folds, the `query_path` comes from it's `base_location`. For non-folds no `fold_path` is needed so the key becomes `(query_path, None)`. This change was needed because the `fold_path` alone is not a unique key.

Looks like we're using the entire `FoldScopeLocation` as the second part of the tuple though, which seems unnecessary to me.

Reply:
I actually think storing this table in the aliases isn't necessary because it is stored inside the SQLFoldObject. The aliases dictionary doesn't ever actually get queried with any tuples for which the second element is non-null.

I am going to remove this redundant book-keeping in this PR. We can discuss adding it again if it is necessary for traversals inside fold (though I think the SQLFoldObject state should be sufficient in that case anyway).